From 663ec037e67b7e466cfe708482bddf8ef6da62c0 Mon Sep 17 00:00:00 2001
From: Nathaniel Kofalt <nathaniel@kofalt.com>
Date: Thu, 12 May 2022 13:25:44 -0500
Subject: [PATCH] cmd/tailscaled: let SOCKS5 listen on a unix socket

Allow the -socks5-server flag to bind to a file instead of a port when
the arg starts with a "file:" prefix.

Fixes #4657

Signed-off-by: Nathaniel Kofalt <nathaniel@kofalt.com>
---
 cmd/tailscaled/tailscaled.go | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/cmd/tailscaled/tailscaled.go b/cmd/tailscaled/tailscaled.go
index fa26ab4d724dc..1733525ee2ff2 100644
--- a/cmd/tailscaled/tailscaled.go
+++ b/cmd/tailscaled/tailscaled.go
@@ -598,9 +598,14 @@ func mustStartProxyListeners(socksAddr, httpAddr string) (socksListener, httpLis
 
 	var err error
 	if socksAddr != "" {
-		socksListener, err = net.Listen("tcp", socksAddr)
+		socksType := "tcp"
+		if strings.HasPrefix(socksAddr, "file:") {
+			socksType = "unix"
+			socksAddr = strings.TrimPrefix(socksAddr, "file:")
+		}
+		socksListener, err = net.Listen(socksType, socksAddr)
 		if err != nil {
-			log.Fatalf("SOCKS5 listener: %v", err)
+			log.Fatalf("SOCKS5 %s listener: %v", socksType, err)
 		}
 		if strings.HasSuffix(socksAddr, ":0") {
 			// Log kernel-selected port number so integration tests
